internal/server/ptypes/workspace.go

```diff
@@ -10,6 +12,25 @@ import (
 	pb "github.com/hashicorp/waypoint/internal/server/gen"
 )
 
+// WorkspaceNameRegexp is the valid Workspace name regular expression.
+// Copied from UsernameRegexp.
+var WorkspaceNameRegexp = regexp.MustCompile(`(?i)^[a-z\d][a-z\d_-]*[a-z\d]+$`)
```

Comment:
This is fine to start with but might be too restrictive cause this is ascii only and not super friendly to foreign languages. We may want to change to unicode character classes at some point: https://www.regular-expressions.info/unicode.html

Reply:
Do we even really need to restrict the workspace name?

Reply:
I would think maybe restrict it to not include spaces, but otherwise I would think not. I borrowed this from Users

Reply:
The easy way to do this is loop over each rune and call `unicode.Is(unicode.Nl, r)`. https://pkg.go.dev/unicode#Is, https://pkg.go.dev/unicode#Nl

Reply:
That's equivalent to the regexp `\p{Nl}+`

Reply:
After some tinkering I updated the regular expression to be `^[\p{L}\p{N}_]*$` in 8bb7e90d13554d5d20bee5f9cdd178a90366a2bb . I used https://regex101.com/r/Ngn8bv/1 to adjust it as I couldn't get Mitchell's suggestion to work as-is, and I don't know regex well enough to know if it was intended as pseudo code or regex that would normally work in some thing other than Golang. 

I also added new tests for this, as well as an additional verification in `workspacePut` where as before we were only verifying in the service.